File: solution/450/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* BSTsearch(TreeNode* root, int key, TreeNode* &nodeparent){
        if(root->val == key)return root;
        else if(root->val > key && root->left){
            nodeparent = root;
            return BSTsearch(root->left, key, nodeparent);
        }else if(root->val < key && root-> right){
            nodeparent = root;
            return BSTsearch(root->right, key, nodeparent);
        }
        return NULL;
    }

    TreeNode* find_successor(TreeNode* node, TreeNode* &nodeparent){
        nodeparent = node;
        TreeNode* ptr = node->right;
        while(ptr->left){
            nodeparent = ptr;
            ptr = ptr->left;
        }
        return ptr;
    }

    void delete_node(TreeNode *sucessor_node, TreeNode *nodeparent){
        if(sucessor_node->val > nodeparent->val){
            if(sucessor_node->left)nodeparent->right = sucessor_node->left;
            else if(sucessor_node->right)nodeparent->right = sucessor_node->right;
            else nodeparent->right = NULL;
        }else{
            if(sucessor_node->left)nodeparent->left = sucessor_node->left;
            else if(sucessor_node->right)nodeparent->left = sucessor_node->right;
            else nodeparent->left = NULL;
        }
    }

    TreeNode* deleteNode(TreeNode* root, int key) {
        if(!root)return root;
        TreeNode* nodeparent = NULL;
        TreeNode* node = BSTsearch(root, key, nodeparent);
        if(!node)return root;
        if(node->left&&node->right){
            TreeNode* successor_node = find_successor(node, nodeparent);
            delete_node(successor_node, nodeparent);
            node->val = successor_node->val;
            return root;
        }
        if(nodeparent){
            delete_node(node, nodeparent);
            return root;
        }
        if(node->left){
            root = root->left;
        }else{
            root = root->right;
        }
        return root;
    }
};

File: solution/450/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def searchnode(self, root, key, parent):
        if root.val == key:return parent, root
        if root.val > key:
            if root.left:
                return self.searchnode(root.left, key, root)
            else:
                return None, None
        else:
            if root.right:
                return self.searchnode(root.right, key, root)
            else:
                return None, None

    def preorder_search(self, node, parent):
        parent = node
        ptr = node.right
        while ptr.left:
            parent = ptr
            ptr = ptr.left
        return parent, ptr

    def nodedelete(self, node, parent):
        if node.val < parent.val:
            if node.left:
                parent.left = node.left
            elif node.right:
                parent.left = node.right
            else:
                parent.left = None
        else:
            if node.right:
                parent.right = node.right
            elif node.left:
                parent.right = node.left
            else:
                parent.right = None

    def deleteNode(self, root, key):
        """
        :type root: TreeNode
        :type key: int
        :rtype: TreeNode
        """
        if not root:return root
        parent, node = self.searchnode(root, key, None)
        if not node:return root
        if node.left and node.right:
            parent, preorder = self.preorder_search(node, parent)
            self.nodedelete(preorder, parent)
            node.val = preorder.val
            return root
        if parent:
            self.nodedelete(node, parent)
            return root
        if node.left:
            return root.left
        else:
            return root.right
